Add Musca-S1 SoC and board support

davidvincze opened the pull request:
Musca-S1 is a Cortex-M33 based SoC. It's similar to the
Musca-B1, but among other things the embedded flash has
been replaced with embedded MRAM (eMRAM) memory.

The Musca-S1 files have been created based on the Musca-B1
SoC and board files.

ioannisg commented:
@davidvincze will you be able to fix CI issues here, before the 2.4 freeze deadline?

microbuilder commented:
This fails when building a TF-M binary, such as via `$ west build -p -b v2m_musca_s1_nonsecure zephyr/samples/tfm_integration/psa_level_1/`

```
Traceback (most recent call last):
  File "/Users/kevin/Linaro/zephyr/upstream/zephyr/../modules/tee/tfm/trusted-firmware-m/bl2/ext/mcuboot/scripts/assemble.py", line 93, in <module>
    main()
  File "/Users/kevin/Linaro/zephyr/upstream/zephyr/../modules/tee/tfm/trusted-firmware-m/bl2/ext/mcuboot/scripts/assemble.py", line 87, in main
    output = Assembly(args.layout, args.output)
  File "/Users/kevin/Linaro/zephyr/upstream/zephyr/../modules/tee/tfm/trusted-firmware-m/bl2/ext/mcuboot/scripts/assemble.py", line 37, in __init__
    self.find_slots()
  File "/Users/kevin/Linaro/zephyr/upstream/zephyr/../modules/tee/tfm/trusted-firmware-m/bl2/ext/mcuboot/scripts/assemble.py", line 48, in find_slots
    offsets = macro_parser.evaluate_macro(self.layout_path, offset_re, 1, 2)
  File "/Users/kevin/Linaro/zephyr/upstream/modules/tee/tfm/trusted-firmware-m/bl2/ext/mcuboot/scripts/macro_parser.py", line 60, in evaluate_macro
    with open(configFile, 'r') as macros_preprocessed_file:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/kevin/Linaro/zephyr/upstream/build/tfm/image_macros_preprocessed.c'
ninja: build stopped: subcommand failed.
```

Recent changes to Zephyr's TF-M module require the following changes from the musca B1 here: https://github.com/zephyrproject-rtos/zephyr/commit/310999da5b6c84f7842c7d797c68da96ebb72f01

microbuilder commented on boards/arm/v2m_musca_s1/CMakeLists.txt:
```diff
@@ -0,0 +1,65 @@
+		set(TFM_IMAGE_VERSION_NS 0.0.0+0)
+	endif()
+
+	set(PREPROCESSED_FILE "${CMAKE_BINARY_DIR}/tfm/image_macros_preprocessed")
```
`PREPROCESSED_FILE` and some other locations have changed in the latest TF-M update, see https://github.com/zephyrproject-rtos/zephyr/commit/310999da5b6c84f7842c7d797c68da96ebb72f01#diff-89eda17a4ad2615d313752b174e955c580154bf123e0d2de83262fa6270e9fae

microbuilder commented on boards/arm/v2m_musca_s1/doc/index.rst:
```diff
@@ -0,0 +1,401 @@
+Building a secure/non-secure with Trusted Firmware
+--------------------------------------------------
+
+The process requires five steps:
```
Some mention of the `v2m_musca_s1_nonsecure` board option should probably be added here as the correct board name to use with TF-M.

davidvincze replied:
Done.

microbuilder commented on boards/arm/v2m_musca_s1/doc/index.rst:
```diff
@@ -0,0 +1,401 @@
+
+In order to build tfm please refer to `Trusted Firmware-M Guide`_.
+Follow the build steps for AN521 target while replacing the platform with
+``-DTARGET_PLATFORM=MUSCA_S1`` and compiler (if required) with ``-DCOMPILER=GNUARM``
```
This seems to be out of date, and should be updated for the recent build system changes around TF-M 1.2.

davidvincze replied:
Done.

davidvincze committed:
arm: V2M Musca-S1: Add Musca-S1 SoC and board support

Musca-S1 is a Cortex-M33 based SoC. It's similar to the
Musca-B1, but among other things the embedded flash has
been replaced with embedded MRAM (eMRAM) memory.

The Musca-S1 files have been created based on the Musca-B1
SoC and board files.

Add the Musca-S1 board to the list of allowed platforms
for the TF-M integration examples.

Change-Id: I4f517d28d0a5b8c4a3fc3fab73adb5519acfc3c2
Signed-off-by: David Vincze <david.vincze@linaro.org>

microbuilder commented:
@ioannisg Any concerns here since this is another TF-M enabled board?

microbuilder commented on boards/arm/v2m_musca_s1/doc/index.rst:
```diff
@@ -0,0 +1,430 @@
+Uploading an application to V2M Musca-S1
+----------------------------------------
+
+To upload the :ref:`hello_world` application to the board no extra steps are
+required, you can upload directly the ``build/zephyr/zephyr.hex`` build
+artefact. However, in general the applications must be converted to Intel's hex
+format before being flashed to a V2M Musca-S1. An optional bootloader can also
+be prepended to the image.
+The eMRAM base address alias is 0xA000000, and the QSPI flash base address alias
+is 0x0. The image offset is calculated by adding the flash offset to the
+bootloader partition size (when there is one).
```
```suggestion
To upload the :ref:`hello_world` application to the board, no extra steps are
required. You can directly upload ``build/zephyr/zephyr.hex``, which is
generated by Zephyr's build system.

In other situations, applications must first be converted to Intel's hex
format before being flashed to a V2M Musca-S1. An optional bootloader can also
be prepended to the image.

The eMRAM base address alias is ``0xA000000``, and the QSPI flash base address
alias is ``0x0``. The image offset is calculated by adding the flash offset to the
bootloader partition size (when there is one).
```

microbuilder commented on boards/arm/v2m_musca_s1/doc/index.rst:
```diff
@@ -0,0 +1,430 @@
+is 0x0. The image offset is calculated by adding the flash offset to the
+bootloader partition size (when there is one).
+
+A third-party tool (srecord) can be used to concatenate the images and generate
+the Intel formatted hex image.
+For more information refer to the `Srecord Manual`_.
```
```suggestion
A third-party tool (srecord) can be used to concatenate the images and generate
the Intel formatted hex image.

For more information refer to the `Srecord Manual`_.
```

microbuilder commented on boards/arm/v2m_musca_s1/doc/index.rst:
```diff
@@ -0,0 +1,430 @@
+   :board: v2m_musca_s1_nonsecure
+   :goals: build
+
+To upload the build artefact to the board first connect the Musca-S1 to your
+computer using the USB port, power it on and copy the ``build/tfm_zephyr.hex``
+file onto the newly appeared MUSCA_S mass storage device (for a more detailed
+description of these steps, please read the 'Uploading an application
+to V2M Musca-S1' section). Once it is done, you may reset the board.
+The ``tfm_zephyr.hex`` file was generated by concatenating the signed TF-M and
+Zephyr binaries with the MCUboot image and converting it to Intel's hex format.
+These steps are all performed automatically by CMake.
```
```suggestion
To upload the build artifact to the board, first connect the Musca-S1 to your
computer using the USB port, power it on and copy the ``build/tfm_zephyr.hex``
file onto the MUSCA_S mass storage device. (For a more detailed
description of these steps, please read the 'Uploading an application
to V2M Musca-S1' section.)

Once the file transfer has completed, you may reset the board.

The ``tfm_zephyr.hex`` file was generated by concatenating the signed TF-M and
Zephyr binaries with the MCUboot image, and converting it to Intel's hex format.
These steps are all performed automatically by CMake.
```

davidvincze committed:
arm: V2M Musca-S1: Update board's index.rst file

Change-Id: I211817620db2130ffb275a5962a24bf90aad57e9
Co-authored-by: Kevin Townsend <kevin@ktownsend.com>
Signed-off-by: David Vincze <david.vincze@linaro.org>

ioannisg reviewed (commented):
Sorry for looking at this too late. 
Looks OK to me, though, thanks for the contribution.
